### PR TITLE
Filter by authors

### PR DESCRIPTION
**Pull request by ailisp**
This PR implements filter post by author in smart contract side in an efficient way. The migration is designed to be scalable.

This is tested by:
- [x] deploy a copy of old contract and ui. Add a couple of posts. at least 3 of two users
- [x] deploy a copy of new contract, call unsafe_add_post_authors & add_old_post_authors to migrate
- [x] check the post is migrated.
- [x] add post with new contract
- [x] check get_posts_by_author works

There are alternative ways to implement this feature, and overall this smart contract side filter is better:
- frontend filter. Similar to hackathon team did in fuzz search. Could be problematic when posts number grow.
- backend filter. need to write and host an indexer

### CODE REVIEW

File: src/lib.rs
Comment (frol) [inline src/lib.rs]: I think in this contract we make an assumption that the number of posts by label and posts by author is not hundred of thousands, so I would just follow the existing pattern of `label_to_posts` and use Rust `HashSet<PostId>` instead of `UnorderedSet<PostId>`.

File: src/migrations.rs
Comment (frol) [inline src/migrations.rs:61]: It is a bit hard to follow the structure of this file when we have two `impl Contract`. I wonder if we can do:

```rust
struct OldContractV1 { ... }

#[near_bindgen]
impl OldContractV1 { }


struct OldContractV2 { ... }

#[near_bindgen]
impl OldContractV2 { }
```

This way we can at least group the code logically.

**Review by frol — APPROVED** [review]
> frontend filter. Similar to hackathon team did in fuzz search. Could be problematic when posts number grow.

Agree.

> backend filter. need to write and host an indexer

This is ideal, but it adds one extra moving piece, so let's postpone this while we can away without going that route.

Overall, this PR is fine (I still left a couple of comments), but we have not figured out the smart contract deployment automation, and whoever will deploy it, must run the migration.

@ailisp Do you have permission to redeploy it in production?

File: src/lib.rs
Comment (ailisp) [reply src/lib.rs]: OK

File: src/migrations.rs
Comment (ailisp) [reply src/migrations.rs:61]: Sounds good!

Comment (ailisp) [reply src/migrations.rs:61]: near_bindgen macro doesn't work on multiple different named struct. I'll name the struct in this way, and impl migrate methods with comments.

**Comment by ailisp** [comment]
@frol all comments addressed and tested again. I don't have permission to production deploy, may you share the key?

**Comment by ailisp** [comment]
Max asked me to add self upgrade and generic migrate method to this contract. With these, we can migrate the contract with a function call access key, and on CD

**Comment by ailisp** [comment]
So i've tested self upgrade and generic migration works:
1. Deploy the old contract (from master) except adding the self_upgrade method:
```rust
// in an impl Contract {
    pub fn unsafe_self_upgrade() {
        near_sdk::assert_self();

        let contract = env::input().expect("No contract code is attached in input");
        Promise::new(env::current_account_id()).deploy_contract(contract);
    }
```
and change this:
```rs
        let mut siblings = self
            .post_to_children
            .get(&parent_id)
            .unwrap_or_else(|| panic!("Parent id {} not found", parent_id));
```
to 
```rs
        let mut siblings = self
            .post_to_children
            .get(&parent_id)
            .unwrap_or_else(|| if id == 0 { Vec::new() } else { panic!("Parent id {} not found", parent_id)});
```
2. add a few posts with different authors, e.g.
```sh
near call c.zxcvn.testnet add_post --accountId zxcvn.testnet --deposit 0.01 --args '{"parent_id":null,"body":{"post_type": "Idea","idea_version":"V1","name":"e","description":"ee"},"labels":[]}'
```
3. go to this branch, call self upgrade with this branch's wasm. e.g.:
```sh
near call c.zxcvn.testnet unsafe_self_upgrade --accountId c.zxcvn.testnet --args $(base64 < res/devgovgigs.wasm ) --base64 --gas 300000000000000
```
4. migrate step 1. e.g.:
```sh
near call c.zxcvn.testnet unsafe_migrate --accountId c.zxcvn.testnet --args '{"V3":"AddPostAuthorsField"}'
```
5. migrate step 2. e.g.:
```sh
near call c.zxcvn.testnet migrate --accountId c.zxcvn.testnet --args '{"to":{"V3":{"InsertPostAuthors":{"start":0,"end":100}}}}'  --gas 300000000000000
# a couple of times in production until all posts are added.
```

7. Add some posts with new contract. You will see posts by authors result is expected. e.g.:
```sh
near view c.zxcvn.testnet get_posts_by_author '{"author":"a.zxcvn.testnet"}'
```

**Comment by ailisp** [comment]
@frol Please take another look and let's figure out a plan to run ^ steps in the CD. My thought:

1. have a PR to just add
```rust
    pub fn unsafe_self_upgrade() {
        near_sdk::assert_self();

        let contract = env::input().expect("No contract code is attached in input");
        Promise::new(env::current_account_id()).deploy_contract(contract);
    }
```

2. manually deploy contract with `unsafe_self_upgrade`
3. add an account `A` in the github secret. add function call access key to enable account `A` to call `unsafe_self_upgrade`, `unsafe_migrate` and `migrate`.
4. In CD, call `unsafe_self_upgrade` for every PR.
5. create this PR specific script, commit, which calls above `unsafe_migrate` and `migrate` with this PR specific args. In CD runs this script.

**Comment by frol** [comment]
@ailisp I suggest we completely eliminate human interactions and implement a migration method that does not need any parameters and return whether we need to call it once again or all the migrations are already applied. Thus, CI will just do the following:

1. Call a deploy function to deploy the code
2. Call a migrate function until it returns that all the migrations are applied (the deploy function may even call the migrate function as a callback once if we have enough gas)

File: src/social_db.rs
Comment (frol) [inline src/social_db.rs]: Can we just drop the `mut` before `data`?

Comment (ailisp) [reply src/social_db.rs]: I think that's not right. The cross contract caller should faithfully follows (ideally import) the interface of callee contract. I looked at https://github.com/NearSocial/social-db/blob/39016e654739b0a3e8cb7ffaea4b03157c4aea6e/contract/src/api.rs#L135, the `mut` cannot be dropped there.

**Comment by ailisp** [comment]
> 

Sounds good! Will try

**Comment by ailisp** [comment]
@frol Now your suggested mechanism works! Please take another look

File: src/migrations.rs
Comment (frol) [inline src/migrations.rs:115]: `new` method have to set the latest state version for it to work properly for the accounts where the contract is deployed with a fresh state.

File: src/migrations.rs
Comment (frol) [inline src/migrations.rs]: Can we batch `unsafe_migrate` in the same receipt with `deploy_contract` or deployment takes too much gas?

File: src/migrations.rs
Comment (frol) [inline src/migrations.rs:150]: **Bikesheding, feel free to ignore it.** I know we want people to "ignore" these methods, but "unsafe" triggers me. Should we just name these methods like `contract_self_deploy` and `contract_self_state_migrate`? I believe "contract_self" will already make these methods "boring enough".

File: src/migrations.rs
Comment (frol) [inline src/migrations.rs]: Is it going to serialize the response with borsh? Let's return a JSON string `"needs-migration"` or `"done"`, so it is explicit. A boolean is hard to extend and reason about.

File: src/social_db.rs
Comment (frol) [reply src/social_db.rs]: `mut` is not part of the interface, unlike `&` or `&mut`, so I would assume it is fine to drop it.

**Review by frol — APPROVED** [review]
I left a few comments that I believe we should address before merging, but I approve the PR to unblock merging after those are addressed.

File: src/migrations.rs
Comment (ailisp) [reply src/migrations.rs:115]: Good point!

File: src/social_db.rs
Comment (ailisp) [reply src/social_db.rs]: That's good to know! I'll drop it then.

File: src/migrations.rs
Comment (ailisp) [reply src/migrations.rs]: I think the help doesn't overweight the problem for these reasons:
- We need call unsafe_migrate a couple of times to do `O(N)` migration task, e.g. add something per post (*).  
- So in general case, we need to call unsafe_self_upgrade once and unsafe_self_migrate multiple times
- Combine unsafe_self_upgrade & unsafe_self_migrate would still check result on the CD script. If it's not done, script would loop call unsafe_self_migrate, which is not simplified compare to not combining unsafe_self_upgrade + unsafe_self_migrate

(*) I measured migrating 100 posts of 2 authors take 57T gas. It's not more than 300T gas yet, but as a generic migration solution, we should plan for when we have 1000 posts and when per post migration action is more expensive than this one.

File: src/migrations.rs
Comment (ailisp) [reply src/migrations.rs:150]: I think `unsafe_*` was mostly to emphasize this method doesn't use `near_bindgen`'s state load deserialization / state serialization and write, parse argument from `env::input`value return, etc. This behavior does not "safely" mutate the state with a near-sdk-rs APIs and implementer should be careful to ensure state consistency

File: src/migrations.rs
Comment (ailisp) [reply src/migrations.rs]: Will do! Yeah that will be easiser to handle on the CD.

File: src/migrations.rs
Comment (frol) [reply src/migrations.rs]: The reason I believe it is beneficial to batch the first migration call together with the deployment is that it will at least migrate the `STATE` value. Without this migration, there is a high chance that view and change calls will fail unable to deserialize the `STATE` for a short period of time after the new code is deployed and the first migration is called.

**Comment by ailisp** [comment]
Comments are addressed. I'll coordinate with @nearmax to do the last manual deploy (to add `self_upgrade` method to production contract, and add access keys), then setup the CD to deploy this into production.

**Comment by ailisp** [comment]
Steps to deploy:
1. Deploy the new contract in the main branch to mainnet. With **near-cli-js**, it is:
```
git checkout main && git pull
export NEAR_ENV=mainnet
near deploy devgovgigs.near res/devgovgigs.wasm
```
2. Call `unsafe_migrate` a few times until all posts are migrated. It will print `done` if so. Otherwise, it prints `needs-migration` and just repeating run this command. It migrates 100 posts a time.
```
near call devgovgigs.near unsafe_migrate --accountId devgovgigs.near --gas 300000000000000
```
3. Add function call access keys to `devgovgigs.near`. So I can setup CD with this key.
```
near add-key devgovgigs.near 5qNj2Kzef8U19ysEFuFgKeTRKSmGrrY8WDYcjd6DB4r9 --contract-id devgovgigs.near --method-names unsafe_self_upgrade --method-names unsafe_migrate --allowance 10000
```

**Comment by maxzaver** [comment]
@ailisp @frol I have executed the above commands.

**Pull request merged**
